trio_asyncio/_base.py

```diff
@@ -91,6 +91,71 @@ def shutdown(self, wait=None):
         self._running = False
 
 
+class AsyncGeneratorDispatcher:
+    """Helper object providing async generator hooks that route
+    finalization to either the correct trio-asyncio event loop or the
+    outer Trio run, depending on where the generator was first iterated.
+    """
+
+    def __init__(self, prev_hooks):
+        self.prev_hooks = prev_hooks
+        self.refcnt = 1
+
+    @classmethod
+    def install(cls):
+        current_hooks = sys.get_asyncgen_hooks()
+
+        # These hooks should either be our own AsyncGeneratorDispatcher
+        # (for another trio-asyncio loop) or Trio's hooks. Both of those
+        # provide both hooks.
+        assert current_hooks.firstiter is not None
+        assert current_hooks.finalizer is not None
+
+        matches = (
+            getattr(current_hooks.firstiter, "__func__", None) is cls.firstiter
+        ) + (getattr(current_hooks.finalizer, "__func__", None) is cls.finalizer)
+        if matches == 0:
+            # Create a new dispatcher that forwards non-trio-asyncio asyncgens
+            # to the current_hooks
+            dispatcher = cls(prev_hooks=current_hooks)
+            sys.set_asyncgen_hooks(
+                firstiter=dispatcher.firstiter, finalizer=dispatcher.finalizer
+            )
+        else:
+            # Take a new reference to the dispatcher that the current_hooks
+            # refer to
+            assert matches == 2
+            dispatcher = current_hooks.firstiter.__self__
+            assert dispatcher is current_hooks.finalizer.__self__
+            assert isinstance(dispatcher, cls)
+            dispatcher.refcnt += 1
+        return dispatcher
+
+    def uninstall(self):
+        self.refcnt -= 1
+        if self.refcnt <= 0:
+            sys.set_asyncgen_hooks(*self.prev_hooks)
+            assert self.refcnt == 0
+
+    def firstiter(self, agen):
+        if sniffio_library.name == "asyncio":
+            loop = asyncio.get_running_loop()
```

Comment:
If there's no running loop, then sniffio should not be reporting that we're in asyncio mode.

Reply:
Python remembers the finalizer hook that was active when an async generator was firstiter'ed, and calls it even if it is no longer the active finalizer hook. Thus, finalizer hooks have to be prepared to be called after the event loop has stopped. This can happen even without our intervention (ie with pure asyncio) so I don't think we need to do anything special. The enqueued task will run if the event loop is resumed.
```
oremanj@shainfra10:/usr/scratch/oremanj$ python3.12
Python 3.12.2 (main, Feb 12 2024, 10:12:13) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> async def agen(): yield
...
>>> async def main():
...   global ag
...   ag = agen()
...   await ag.asend(None)
...
>>> loop = asyncio.new_event_loop()
>>> loop.run_until_complete(main())
>>>

# now we have a stopped event loop and a half-finished async generator:
>>> loop._asyncgens
{<weakref at 0x7f7004348b80; to 'async_generator' at 0x7f700472eda0 (agen)>}
>>> loop._ready
deque([])

# if we destroy the async generator, asyncio will schedule its finalizer:
>>> del ag
>>> loop._asyncgens
set()
>>> loop._ready
deque([<Handle BaseEventLoop.create_task(<async_genera...x7f700432fc00>)>])
```